core/src/main/java/com/linecorp/armeria/common/util/DefaultLimitedBlockingTaskExecutor.java

```diff
@@ -0,0 +1,240 @@
+            } finally {
+                taskCounter.decrementAndGet();
+            }
+        }, delay, unit);
+    }
+
+    @Override
+    public ScheduledFuture<?> scheduleAtFixedRate(Runnable command, long initialDelay, long period,
+                                                  TimeUnit unit) {
+        return delegate.scheduleAtFixedRate(() -> {
+            try {
+                taskCounter.incrementAndGet();
+                command.run();
+            } finally {
+                taskCounter.decrementAndGet();
+            }
+        }, initialDelay, period, unit);
+    }
+
+    @Override
+    public ScheduledFuture<?> scheduleWithFixedDelay(Runnable command, long initialDelay, long delay,
+                                                     TimeUnit unit) {
+        return delegate.scheduleAtFixedRate(() -> {
+            try {
+                taskCounter.incrementAndGet();
+                command.run();
+            } finally {
+                taskCounter.decrementAndGet();
+            }
+        }, initialDelay, delay, unit);
+    }
+
+    @Override
+    public void shutdown() {
+        delegate.shutdown();
+    }
+
+    @Override
+    public List<Runnable> shutdownNow() {
+        return delegate.shutdownNow();
+    }
+
+    @Override
+    public boolean isShutdown() {
+        return false;
+    }
+
+    @Override
+    public boolean isTerminated() {
+        return delegate.isShutdown();
+    }
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+        return delegate.awaitTermination(timeout, unit);
+    }
+
+    @Override
+    public <T> Future<T> submit(Callable<T> task) {
+        return delegate.submit(() -> {
+            try {
+                taskCounter.incrementAndGet();
+                return task.call();
+            } finally {
+                taskCounter.decrementAndGet();
+            }
+        });
+    }
+
+    @Override
+    public <T> Future<T> submit(Runnable task, T result) {
+        return delegate.submit(() -> {
+            try {
+                taskCounter.incrementAndGet();
+                task.run();
+            } finally {
+                taskCounter.decrementAndGet();
+            }
+        }, result);
+    }
+
+    @Override
+    public Future<?> submit(Runnable task) {
+        return delegate.submit(() -> {
+            try {
+                taskCounter.incrementAndGet();
+                task.run();
+            } finally {
+                taskCounter.decrementAndGet();
+            }
+        });
+    }
+
+    @Override
+    public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks) throws InterruptedException {
+        return delegate.invokeAll(tasks.stream()
+                                       .map(task -> (Callable<T>) () -> {
+                                           try {
+                                               taskCounter.incrementAndGet();
```

Comment:
I think we don't need to increase by 1 here. i.e.
```java
newTasks = tasks
  .stream()
  .map(...)
  .collect(...);
final int numTasks = newTasks.size();
taskCounter.addAndGet(numTasks);
try {
  ... decrement by 1 on each task's completion ...
} finally {
  if (!submitted) {
    taskCounter.addAndGet(-numTasks);
  }
}
```

Reply:
I thought a little about it. To check `submitted` for each tasks, I think to need to use `Map<Task, Boolean>`  for check each task. How do you think about it?.?

```java
final Map<Task, Boolean> submits = tasks.stream().collect();
final int numTasks = submits.size();
taskCounter.addAndGet(numTasks);

try {
... remove submit element on each task's completion ...
}  finally {
   taskCounter.addAndGet(-submits.size());
}
```